File: algorithms/python/198_house_robber/house_robber.py
"""
198. House Robber

打家劫舍
"""

from typing import List


def rob(nums: List[int]) -> int:
    """
    动态规划
    """
    prev_max, cur_max = 0, 0
    for i in range(len(nums)):
        temp = cur_max
        cur_max = max(cur_max, prev_max + nums[i])
        prev_max = temp

    return cur_max

File: algorithms/198-house-robber/house-robber.go
package problem198

func rob(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	n := len(nums)
	if len(nums) == 1 {
		return nums[0]
	}
	first := nums[0]
	second := max(nums[0], nums[1])
	for i := 2; i < n; i++ {
		first, second = second, max(first + nums[i], second)
	}
	
	return second
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}